### DISCUSSION TITLE
setting a timer for a lamp

### DISCUSSION THREAD

**Discussion by Pietschnitzel** [jasonacox/tinytuya#316] [Q&A]
when using the app, you have the ability to  set a schedule, in the device logs it shows up as shown here
![image](https://user-images.githubusercontent.com/63900327/229114523-025380b4-30e0-45cf-a107-6aea7814ee7e.png)
i would want to add that scheduling ability to my Programm, but i cannot find out how the function works, especially the dps_id parameter

Any help is appreciated

**Comment by uzlonewolf** [comment]
Are you referring to a recurring timer, or the 1-shot countdown timer?  Very few devices can do recurring timers on-device, for most of them the timer is actually in the cloud and won't function without internet.  1-shot countdown timers should be on-device though and can be set if you know the right DP.

If you start a persistent monitoring loop and change the timer with the app, do you see a DP update?
```python
import time
import tinytuya

d = tinytuya.Device( ..., persist=True )

print( d.status() )
print(" > Begin Monitor Loop <")
pingtime = time.time() + 9
polltime = time.time() + 300

while(True):
    if( pingtime <= time.time() ):
        payload = d.generate_payload(tinytuya.HEART_BEAT)
        data = d.send(payload)
        pingtime = time.time() + 9
        if data:
            print( data )

    data = d.receive()
    if data:
        print( data )
```

**Reply by Pietschnitzel** [reply]
the programm doesnt show anything, but the logs do
![image](https://user-images.githubusercontent.com/63900327/229138474-3f12c930-30f9-444f-a824-e9fbad685eda.png)
![image](https://user-images.githubusercontent.com/63900327/229138620-656e628a-d869-4dd9-9c91-4db3193308cc.png)

**Reply by jasonacox** [reply]
Interesting.  Looks like a cloud only function. Do you see anything with the timertest.py script when you turn the lamp on and off?

**Reply by Pietschnitzel** [reply]
yep, 
`{'devId': 'ID', 'dps': {'1': True}, 't': 1680368587, 's': 10}` and
`{'devId': 'ID', 'dps': {'1': False}, 't': 1680368589, 's': 11}`

it might be cause i used the wrong arguments wrong?
cause if the logs show it correctly.
can one of you send an example of how the function with argument should look when trying to turn on default white

**Reply by jasonacox** [reply]
Happy to take a look at your code if you want to paste it.

> can one of you send an example of how the function with argument should look when trying to turn on default white

Here is a examples script to set white and rotates through the rainbow:

```python
import tinytuya
import time

d = tinytuya.BulbDevice(DEVICEID, DEVICEIP, DEVICEKEY)
d.set_version(3.3)
d.set_socketPersistent(True)  

# Show status of device
data = d.status()
print('\nCurrent Status of Bulb: %r' % data)

# Set to full brightness warm white
print('\White Test')
d.set_white()

# Flip through colors of rainbow - set_colour(r, g, b):
print('\nColor Test - Cycle through rainbow')
rainbow = {"red": [255, 0, 0], "orange": [255, 127, 0], "yellow": [255, 200, 0],
           "green": [0, 255, 0], "blue": [0, 0, 255], "indigo": [46, 43, 95],
           "violet": [139, 0, 255]}
for x in range(2):
    for i in rainbow:
        r = rainbow[i][0]
        g = rainbow[i][1]
        b = rainbow[i][2]
        print('    %s (%d,%d,%d)' % (i, r, g, b))
        d.set_colour(r, g, b)
        time.sleep(WAIT)
    print('')

# Done
print('\nDone')
d.set_white()
```
